src/cascadia/WindowsTerminal/AppHost.cpp

```diff
@@ -92,9 +92,13 @@ AppHost::AppHost() noexcept :
 
 AppHost::~AppHost()
 {
-    if (_window->IsQuakeWindow())
+    if (_windowManager.IsMonarch())
     {
-        _windowManager.RequestHideTrayIcon();
+        _DestroyTrayIcon();
+    }
+    else if (_window->IsQuakeWindow())
+    {
+        _HideTrayIconRequested();

```

Comment:
This is a bad idea; `_HideTrayIconRequested` is a coroutine that takes an implicit reference to `this`, and you're dispatching it out of the destructor. By the time the `resume_backround` resumes you, `this` will have been destroyed.

This needs to be restructured such that (1) the destructor never uses anything that requires a living `this` to persist after the destructor returns and (2) it is as small as possible :smile:

(Also, there's no guarantee that AppHost *will* be destructed.)

Reply:
> Also, there's no guarantee that AppHost will be destructed.

It's a regular instance in `wWinMain`. It should be destroyed properly unless the app `abort()`s, right?
We could make a `_HideTrayIconRequestedAsync` and have `_HideTrayIconRequested` run synchronously.
Also: `_HideTrayIconRequested` accesses settings from a background thread which is a race condition... I should add thread asserts to CascadiaSettings. 🤔

Reply:
> It's a regular instance in `wWinMain`. It should be destroyed properly unless the app `abort()`s, right?

Yes. That does assume that main exits normally. If we were more like a standard UWP Xaml application, we would actually ExitProcess/TerminateProcess when the window was destroyed. The fact that we do _not_ makes us uncommon in the world of Xaml, **and** is the cause for some of our "this object was already disposed of!" error traces on shutdown in debug mode. :smile:

(Because of that, we've entertained the idea multiple times of switching -- it would clear up almost all of our on-exit crashes that originate deep inside Windows.UI.Xaml.dll!)

Reply:
Hmm then I guess an alternative might be to do this particular Tray Icon cleanup in something like `LastTabClosed`?